src/main/java/com/google/googleidentity/oauth2/authorizationcode/AuthorizationCodeRequestHandler.java

```diff
@@ -0,0 +1,118 @@
+/*
+    Copyright 2020 Google LLC
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+
+package com.google.googleidentity.oauth2.authorizationcode;
+
+import com.google.googleidentity.oauth2.exception.InvalidGrantException;
+import com.google.googleidentity.oauth2.exception.OAuth2Exception;
+import com.google.googleidentity.oauth2.request.OAuth2Request;
+import com.google.googleidentity.oauth2.request.RequestHandler;
+import com.google.googleidentity.oauth2.token.OAuth2AccessToken;
+import com.google.googleidentity.oauth2.token.OAuth2TokenService;
+import com.google.googleidentity.oauth2.util.OAuth2Constants;
+import com.google.googleidentity.oauth2.util.OAuth2ParameterNames;
+import com.google.inject.Inject;
+import net.minidev.json.JSONObject;
+import org.apache.http.client.utils.URIBuilder;
+
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.time.Instant;
+import java.util.Optional;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * An Implementation of {@link RequestHandler}, handle request in Authorization Code Flow.
+ */
+public class AuthorizationCodeRequestHandler implements RequestHandler {
+
+    private final AuthorizationCodeService authorizationCodeService;
+
+    private final Logger log = Logger.getLogger("AuthorizationCodeRequestHandler");
+
+    private final OAuth2TokenService oauth2TokenService;
+
+    @Inject
+    public AuthorizationCodeRequestHandler(
+            AuthorizationCodeService authorizationCodeService,
+            OAuth2TokenService oauth2TokenService) {
+        this.authorizationCodeService = authorizationCodeService;
+        this.oauth2TokenService = oauth2TokenService;
+    }
+
+    @Override
+    public void handle(HttpServletResponse response, OAuth2Request oauth2Request)
```

Comment:
What endpoint is this implementing?

You're branching on `response_type` which makes me think this is the authorization endpoint. The `response_type=code` case also looks correct to me.

But, the `response_type=token` case looks like it is implementing the token endpoint logic for authorization code exchange. That's not what `response_type=token` is supposed to do on the authorization endpoint. For `response_type=token`, the user has signed in, given consent to issue a token, and now we're redirecting back to the RP with an access token. In this case, there shouldn't be an auth code issued at any point, nor should a refresh token be returned. See https://tools.ietf.org/html/rfc6749#section-4.2

Reply:
This situation will be handled by ImplicitRequestHandler that I will add later. For authorization code flow, I think this handler is right.